### ISSUE TITLE
Symbol casting to Enums does not work in subclass initializers

### ISSUE THREAD

**Issue by petr-fischer** [crystal-lang/crystal#14370]
In code below, last line does not compile:

[Playground here](https://play.crystal-lang.org/#/r/gl7r).

```crystal
enum MyEnum
  Blah
  Bleh
end
  
class A
  property en : MyEnum
 
  def initialize(@en)
  end
end
 
class B < A
end
 
class C < A
  property name : String
  
  # new initializer here
  def initialize(@en, @name)
  end
end
 
p A.new(:blah).en.to_s
p B.new(:blah).en.to_s
p C.new(MyEnum::Blah, "name").en.to_s
p C.new(:blah, "name").en.to_s # <= Error: instance variable '@en' of A must be MyEnum, not Symbol
```

In the classes `A` and `B`, Symbol casting to Enum works, but not in the class `C`.

**Comment by Blacksmoke16** [comment]
Don't quote me on this, but I'm guessing that the compiler is doesn't really have an explicit type on the `@en` var as it would be inherited from the parent type. Symbol to enum autocasting requires the explicit type restrictions. I.e. if you do `@en : MyEnum` on the child initializer, it'll work fine.

So this may be more of a feature/enhancement request versus an actual bug?

**Comment by devnote-dev** [comment]
I'd say it's more a bug because I don't know why you'd want this as a feature, and instance variable inheritance is confusing enough as it is...

**Comment by Blacksmoke16** [comment]
I guess you could argue it both ways. It's not a bug because this feature requires explicit type restrictions to work, but an enhancement could be to make it smarter in cases like this. But in either case it's usually a good practice to use type restrictions, so then you could also argue this is a case of "just be explicit and it works" or at least low on the priority list.

**Comment by ysbaddaden** [comment]
This is two-fold:

1. casting a symbol to an enum requires the argument's type to be explicit (`value : MyEnum`).
2. `def method(@value); end` gets transformed to `def method(value) @value = value; end` in which case the `value` argument becomes typeless.

Now, a feature enhancement could be for the ivar expansion in methods to automatically include the ivar type. I'd love it, though I'm not sure how easy or hard it would be to implement, especially if we expand the syntax before we start inferring the ivar types (we don't want to bring event more confusion by having some ivar arguments typed while others are not.

**Comment by straight-shoota** [comment]
I don't think it should be too difficult. When expanding an ivar paramter name, we can copy the ivar's type restriction to the parameter. That should make it work.
